### PR TITLE
Explicitly signal stream initiation and termination

### PR DESCRIPTION
**Pull request by NickCao**

### CODE REVIEW

File: jumpstarter/drivers/base.py [hunk 1]
```diff
@@ -98,12 +98,14 @@ async def Stream(self, request_iterator, context):
...
                 await resource.send(str(resource_uuid).encode("utf-8"))
+                await resource.send_eof()
 
                 async with remote:
                     async for v in forward_server_stream(request_iterator, remote):
                         yield v
 
-                del self.resources[resource_uuid]
+                # del self.resources[resource_uuid]
+                # small resources might be fully buffered in memory
```
Comment (mangelajo) [inline jumpstarter/drivers/base.py:108]: What do you mean by this?

File: jumpstarter/common/streams.py [hunk 2]
```diff
@@ -1,58 +1,70 @@
...
+                case router_pb2.FRAME_TYPE_DATA:
+                    await tx.send(frame.payload)
+                case router_pb2.FRAME_TYPE_GOAWAY:
+                    if isinstance(tx, ObjectStream) or isinstance(tx, ByteStream):
+                        await tx.send_eof()
+                case _:
+                    pass
+    # ignore peer disconnect
+    except BrokenResourceError:
+        pass
```
Comment (mangelajo) [inline jumpstarter/common/streams.py:34]: we are doing excessive use of pass , in some cases we should at least log it (warning), which will help in the future if issues arise from it.

File: jumpstarter/common/streams.py [hunk 3]
```diff
@@ -1,58 +1,70 @@
...
+    # ignore rpc cancellation and internal error
+    except grpc.aio.AioRpcError as e:
+        match e.code():
+            case grpc.StatusCode.CANCELLED | grpc.StatusCode.INTERNAL:
                 pass
```
Comment (mangelajo) [inline jumpstarter/common/streams.py:39]: same here.

File: jumpstarter/common/streams.py [hunk 4]
```diff
@@ -1,58 +1,70 @@
...
+    try:
+        async for frame in rx:
+            match frame.frame_type:
+                case router_pb2.FRAME_TYPE_DATA:
+                    await tx.send(frame.payload)
+                case router_pb2.FRAME_TYPE_GOAWAY:
+                    if isinstance(tx, ObjectStream) or isinstance(tx, ByteStream):
+                        await tx.send_eof()
+                case _:
+                    pass
```
Comment (mangelajo) [inline jumpstarter/common/streams.py:31]: see my comments on pass.

**Review by mangelajo — CHANGES_REQUESTED** [review]
Looks great, just some comments about logging all "pass calls"

File: jumpstarter/drivers/base.py (hunk 1 above)
Comment (NickCao) [reply jumpstarter/drivers/base.py:108]: Sequence of event:
1. client opens a stream for sharing resource
2. exporter accepts the stream and registers the other end of the stream and a resource
3. exporter starts task to forward data from the client side of the stream to the exporter side of the stream
3.1 (actually happened) stream is copied into internal buffer before being consumed, forward tasks exits, resource unregistered
4. driver get the stream from the dict and begins consuming the stream
4.1 (actually happened) driver fails to find resource
6. (originally intended) stream is fully consumed by the driver, forward tasks exits, resource unregistered

Comment (NickCao) [reply jumpstarter/drivers/base.py:108]: TLDR: resource shouldn't be unregistered when forwarding task exits, but only when it's fully consumed (or client decides to cancel)

File: jumpstarter/common/streams.py (hunk 2 above)
Comment (NickCao) [reply jumpstarter/common/streams.py]: Opted for debug level, warning is too strong for these (expected to happen in daily use) errors.

File: jumpstarter/drivers/base.py (hunk 1 above)
Comment (mangelajo) [reply jumpstarter/drivers/base.py:108]: Thanks for the explanation I see now :-)

**Review by mangelajo — APPROVED** [review]

**Pull request merged**